frontends/p4/staticAssert.h

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2022 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _FRONTENDS_P4_STATICASSERT_H_
+#define _FRONTENDS_P4_STATICASSERT_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/methodInstance.h"
+
+namespace P4 {
+
+/**
+ * Evaluates static_assert invocations.
+ */
+class DoStaticAssert : public Inspector {
+    ReferenceMap* refMap;
+    TypeMap* typeMap;
+
+ public:
+    DoStaticAssert(ReferenceMap* refMap, TypeMap* typeMap):
+            refMap(refMap), typeMap(typeMap)
+    { CHECK_NULL(refMap); CHECK_NULL(typeMap); setName("DoStaticAssert"); }
+    bool preorder(const IR::MethodCallExpression* method) override {
+        MethodInstance* mi = MethodInstance::resolve(method, refMap, typeMap);
+        if (auto ef = mi->to<ExternFunction>()) {
+            if (ef->method->name == "static_assert") {
+                auto subst = ef->substitution;
+                auto params = subst.getParametersInOrder();
+                if (!params->moveNext()) {
+                    ::warning(ErrorType::WARN_INVALID,
+                              "static_assert with no arguments: %1%", method);
+                    return false;
+                }
+                auto param = params->getCurrent();
+                CHECK_NULL(param);
+                auto arg = subst.lookup(param);
+                CHECK_NULL(arg);
+                if (auto bl = arg->expression->to<IR::BoolLiteral>()) {
+                    if (!bl->value) {
+                        cstring message = "static_assert failed";
+                        if (params->moveNext()) {
+                            param = params->getCurrent();
+                            CHECK_NULL(param);
+                            auto msg = subst.lookup(param);
+                            CHECK_NULL(msg);
+                            if (auto sl = msg->expression->to<IR::StringLiteral>()) {
+                                message = sl->value;
+                            }
+                        }
+                        ::error(ErrorType::ERR_EXPECTED, "%1%: %2%",
+                                method, message);
+                    }
+                } else {
+                    ::warning("Could not evaluate static_assert to a constant: %1%", arg);
```

Comment:
This is the behavior we want? A static assert that cannot be statically resolved is just a warning and not an error.

Reply:
I actually expect that this will fail in typechecking, since the parameters are directionless.
So this case should never arise in practice.

Reply:
I guess this would signal a failure of constant-folding some expression.

Reply:
Ahh right. Makes sense.

Reply:
Should this be a `BUG` then?

Reply:
It could be a BUG as well, if we are confident in the constant-folding code.
I say we leave it like this for now until someone complains about hitting the warning.